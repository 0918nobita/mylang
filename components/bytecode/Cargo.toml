[package]
name = "bytecode"
version = "0.1.0"
edition = "2021"

[dependencies]
anyhow = "1.0.53"
bincode = "1.3.3"
serde = { version = "1.0.136", features = ["derive"] }
