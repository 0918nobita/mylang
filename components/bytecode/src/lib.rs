use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub enum Inst {
    I32Const(i32),
    I32Add,
    Print,
}
