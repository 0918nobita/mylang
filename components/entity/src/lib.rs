use std::fmt::{Debug, Formatter, Result as FmtResult};

#[derive(PartialEq)]
pub struct I32Entity(i32);

impl I32Entity {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn add(&self, rhs: &Self) -> Self {
        Self(self.0 + rhs.0)
    }
}

impl Debug for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

#[derive(PartialEq)]
pub struct StrEntity(String);

impl StrEntity {
    pub fn new(value: String) -> Self {
        Self(value)
    }
}

impl Debug for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

#[derive(Debug)]
pub enum Entity {
    I32(I32Entity),
    Str(StrEntity),
}
