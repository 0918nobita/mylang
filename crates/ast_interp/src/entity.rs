//! 抽象構文木インタプリタで扱う「実体」の定義

use std::fmt::{Debug, Display, Formatter, Result as FmtResult};

/// 符号付き32ビット整数値
pub struct I32Entity(i32);

impl I32Entity {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn add(&self, rhs: &Self) -> Self {
        Self(self.0 + rhs.0)
    }
}

impl Debug for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// 文字列データ
pub struct StrEntity(String);

impl StrEntity {
    pub fn new(value: String) -> Self {
        Self(value)
    }
}

impl Debug for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// 実行時型情報
#[derive(Debug)]
pub enum RuntimeTypeInfo {
    I32,
    Str,
}

impl Display for RuntimeTypeInfo {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(
            f,
            "{}",
            match self {
                RuntimeTypeInfo::I32 => "i32",
                RuntimeTypeInfo::Str => "str",
            }
        )
    }
}

/// バイトコード実行時に扱う値
#[derive(Debug)]
pub enum Entity {
    I32(I32Entity),
    Str(StrEntity),
}

impl Entity {
    pub fn get_type(&self) -> RuntimeTypeInfo {
        match self {
            Entity::I32(_) => RuntimeTypeInfo::I32,
            Entity::Str(_) => RuntimeTypeInfo::Str,
        }
    }
}
