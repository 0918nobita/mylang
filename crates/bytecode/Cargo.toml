[package]
name = "mylang_bytecode"
version = "0.1.0"
edition = "2021"

[dependencies]
serde = { version = "1.0.156", features = ["derive"] }
