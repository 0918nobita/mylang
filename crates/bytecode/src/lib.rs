//! バイトコードの定義

use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub enum Inst {
    I32Const(i32),
    I32Add,
    StrConst(String),
    PrintI32,
    PrintStr,
}
