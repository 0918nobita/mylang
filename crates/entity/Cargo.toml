[package]
name = "mylang_entity"
version = "0.1.0"
edition = "2021"

[dependencies]
