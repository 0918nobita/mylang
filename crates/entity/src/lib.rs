use std::fmt::{Debug, Display, Formatter, Result as FmtResult};

use serde::Serialize;

/// 符号付き32ビット整数値
#[derive(PartialEq, Serialize)]
pub struct I32Entity(i32);

impl I32Entity {
    pub fn new(value: i32) -> Self {
        Self(value)
    }

    pub fn add(&self, rhs: &Self) -> Self {
        Self(self.0 + rhs.0)
    }
}

impl Debug for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for I32Entity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// 文字列データ
#[derive(PartialEq, Serialize)]
pub struct StrEntity(String);

impl StrEntity {
    pub fn new(value: String) -> Self {
        Self(value)
    }
}

impl Debug for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

impl Display for StrEntity {
    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
        write!(f, "{}", self.0)
    }
}

/// バイトコード実行時に扱う値
#[derive(Debug, Serialize)]
pub enum Entity {
    I32(I32Entity),
    Str(StrEntity),
}
