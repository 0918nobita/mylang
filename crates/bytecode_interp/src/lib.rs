//! バイトコードインタプリタ

use bytecode::Inst;
use entity::{Entity, I32Entity, RuntimeTypeInfo, StrEntity};
use thiserror::Error;

/// バイトコードインタプリタで発生するエラー
#[derive(Debug, Error)]
pub enum InterpError {
    /// スタックが空であり、値をスタックから取り出せなかった
    #[error("Stack underflow ({0})")]
    StackUnderflow(String),

    /// 取り出した値が、期待した型の値ではなかった
    #[error("Type mismatch (expected: {expected:?}, actual: {actual:?})")]
    TypeMismatch {
        expected: RuntimeTypeInfo,
        actual: RuntimeTypeInfo,
    },
}

/// バイトコードの実行結果
type InterpResult<T> = Result<T, InterpError>;

/// I32Const 命令を実行する
fn i32_const(stack: &mut Vec<Entity>, immediate: i32) {
    stack.push(Entity::I32(I32Entity::new(immediate)));
}

/// I32Add 命令を実行する
fn i32_add(stack: &mut Vec<Entity>) -> InterpResult<()> {
    let lhs = stack.pop().ok_or_else(|| {
        InterpError::StackUnderflow("Failed to get left-hand side of the addition".to_owned())
    })?;
    let rhs = stack.pop().ok_or_else(|| {
        InterpError::StackUnderflow("Failed to get right-hand side of the addition".to_owned())
    })?;

    let lhs = match lhs {
        Entity::I32(i32_entity) => i32_entity,
        _ => {
            return Err(InterpError::TypeMismatch {
                expected: RuntimeTypeInfo::I32,
                actual: lhs.get_type(),
            })
        }
    };

    let rhs = match rhs {
        Entity::I32(i32_entity) => i32_entity,
        _ => {
            return Err(InterpError::TypeMismatch {
                expected: RuntimeTypeInfo::I32,
                actual: rhs.get_type(),
            })
        }
    };

    stack.push(Entity::I32(lhs.add(&rhs)));
    Ok(())
}

/// StrConst 命令を実行する
fn str_const(stack: &mut Vec<Entity>, immediate: &str) {
    stack.push(Entity::Str(StrEntity::new(immediate.to_owned())));
}

/// PrintI32 命令を実行する
fn print_i32(stack: &mut Vec<Entity>) -> InterpResult<()> {
    let ent = stack.pop().ok_or_else(|| {
        InterpError::StackUnderflow("Failed to get the entity to output".to_owned())
    })?;

    let ent = match ent {
        Entity::I32(i32_entity) => i32_entity,
        _ => {
            return Err(InterpError::TypeMismatch {
                expected: RuntimeTypeInfo::I32,
                actual: ent.get_type(),
            })
        }
    };

    println!("{}", ent);
    Ok(())
}

/// PrintStr 命令を実行する
fn print_str(stack: &mut Vec<Entity>) -> InterpResult<()> {
    let ent = stack
        .pop()
        .ok_or_else(|| InterpError::StackUnderflow("Failed to get entity".to_owned()))?;

    let ent = match ent {
        Entity::Str(str_entity) => str_entity,
        _ => {
            return Err(InterpError::TypeMismatch {
                expected: RuntimeTypeInfo::Str,
                actual: ent.get_type(),
            })
        }
    };

    println!("{}", ent);
    Ok(())
}

/// バイトコードを解釈実行する
pub fn execute(insts: impl Iterator<Item = Inst>) -> InterpResult<()> {
    let mut stack = Vec::<Entity>::new();

    for inst in insts {
        match inst {
            Inst::I32Const(i) => i32_const(&mut stack, i),

            Inst::I32Add => i32_add(&mut stack)?,

            Inst::StrConst(s) => str_const(&mut stack, &s),

            Inst::PrintI32 => print_i32(&mut stack)?,

            Inst::PrintStr => print_str(&mut stack)?,
        }
    }

    Ok(())
}

#[cfg(test)]
mod tests {
    use entity::{Entity, I32Entity, StrEntity};

    use crate::i32_add;

    #[test]
    fn test_i32_add() {
        let mut stack = vec![
            Entity::I32(I32Entity::new(3)),
            Entity::I32(I32Entity::new(4)),
        ];
        insta::assert_debug_snapshot!((i32_add(&mut stack), stack));
    }

    #[test]
    fn test_i32_add_underflow() {
        let mut stack = vec![Entity::I32(I32Entity::new(3))];
        insta::assert_debug_snapshot!((i32_add(&mut stack), stack));
    }

    #[test]
    fn test_i32_add_type_mismatch() {
        let mut stack = vec![
            Entity::Str(StrEntity::new("foo".to_owned())),
            Entity::I32(I32Entity::new(4)),
        ];
        insta::assert_debug_snapshot!((i32_add(&mut stack), stack));
    }
}
