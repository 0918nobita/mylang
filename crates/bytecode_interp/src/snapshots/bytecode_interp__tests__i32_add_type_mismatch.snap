---
source: crates/bytecode_interp/src/lib.rs
assertion_line: 113
expression: "(i32_add(&mut stack), stack)"

---
(
    Err(
        TypeMismatch,
    ),
    [],
)
